File: dynamic_closure/customizations/bom/bom.py
from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.model.document import Document
from erpnext.controllers.item_variant import create_variant
from six import string_types
import json


@frappe.whitelist()
def get_items(doctype, txt, searchfield, start, page_len, filters, as_dict):
	return frappe.db.sql(""" 
		select 
			name 
		from 
			`tabItem` 
		where 
			is_sales_item = 1 
			and has_variants < 2
			and name like '{txt}' """.format(txt= "%%%s%%" % txt))

@frappe.whitelist()
def get_boms(item):
	bom = frappe.get_value("BOM",{"item":item,"is_default":1,"is_active":1},["name","bom_level"],as_dict=1)
	if bom:
		return bom

@frappe.whitelist()
def get_attribute_value(doctype, txt, searchfield, start, page_len, filters, as_dict):
	return frappe.db.sql(""" 
		select 
			name 
		from 
			`tabItem Attribute Value` 
		where 
			parent = '{parent}'
			and name like '{txt}' """.format(parent= filters.get('attribute'),txt= "%%%s%%" % txt))

File: dynamic_closure/customizations/bom/bom.js

frappe.ui.form.on("BOM", {
	validate: function(frm) {
		if(frm.doc.items && frm.doc.uom == "Inch")
		{
			$.each(frm.doc.items, function(i, v) {
				frappe.model.set_value(v.doctype, v.name, "uom", "Feet");
				if(!v.pieces)
				{
					v.pieces = 1;
				}
				var qty = v.pieces*frm.doc.quantity/12;
				frappe.model.set_value(v.doctype, v.name, "qty", qty)
			})
		}
	},
	refresh: function(frm) {
		frm.fields_dict['items'].grid.get_field("item_code").get_query = function(doc, cdt, cdn) {
			return {
				query: "dynamic_closure.customizations.bom.bom.get_items",
				filters: {'is_sales_item': 1}
			}
		}
		frm.fields_dict['items'].grid.get_field("related_attribute_value").get_query = function(doc, cdt, cdn) {
			return {
				query: "dynamic_closure.customizations.bom.bom.get_attribute_value",
				filters: {'attribute': frm.doc.attribute}
			}
		}
	}
});

frappe.ui.form.on("BOM Item", {
	item_code: function(frm, cdt, cdn) {
		var row = locals[cdt][cdn]
		if(row.item_code)
		{
			if(frm.doc.uom == "Inch")
			{
				if(!row.pieces)
				{
					row.pieces = 1;
				}
				var qty = row.pieces*frm.doc.quantity/12;
				frappe.model.set_value(cdt, cdn, "uom", "Feet")
				frappe.model.set_value(cdt, cdn, "qty", qty)
			}
			frappe.call({
				method: "dynamic_closure.customizations.bom.bom.get_boms",
				args:{
					"item": row.item_code
				},
				callback: (r) => {
					if(r.message)
					{
						frm.set_value("bom_level",r.message.bom_level + 1);
						frappe.model.set_value(cdt, cdn, "bom", r.message.name)
					}
					else
					{
						frm.set_value("bom_level",0);
					}

				}
			});
		}
	}
});